Adaptive serving with Network Information API

mihajlija opened the pull request:
Fixes #665 .

---
This pull request adds a **new**:
- [x ] Guide 📘
- [x ] Codelab 🔬

Have you read the wiki on writing [a guide](https://github.com/GoogleChrome/web.dev/wiki/How-to-write-a-guide) or [a codelab](https://github.com/GoogleChrome/web.dev/wiki/How-to-write-a-codelab)?
- [x ] Done! 🍰

Does your content tell a user how to fix a Lighthouse audit?
Have you added [the Lighthouse audit ID to your YAML](https://github.com/GoogleChrome/web.dev/wiki/How-to-write-a-guide#lighthouse-audits)?
- [x ] My changes do not mention Lighthouse audits. 🙅‍♂️

Does your content reference any Glitch projects?
Have you read the wiki on [working with Glitch](https://github.com/GoogleChrome/web.dev/wiki/Working-with-Glitch)?
Have you added the Glitch project to [our team page](https://glitch.com/@webdev)?
- [x ] I have read the wiki on working with Glitch. 🍰
- [x ] I have added my Glitches to the team page. 👏

Meggin commented:
Quick question-- would this content belong a bit better in Network Resilience narrative?

mihajlija commented:
I fits there too, I think it's up to @addyosmani to decide :) The current plan is to group this with Save Data and Clint Hints docs in "Optimize how you serve content" section of /fast.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+title: Adaptive serving based on network quality
+author: mihajlija
+description: |
+  Using Network Information API to adapt which assets you're serving to users based on the quality of their connection.
```
Slight rewrite of this sentence: "Use Network Information API to adapt the assets served to users based on the quality of their connection."

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+
+# Adaptive serving based on network quality
+
+Loading a website can be a very different experience depending on the network conditions. Everything is usually smooth when you are on a fast network, but when you are on the go with a humble data plan and spotty connection, or stuck with a laptop on a sluggish coffee shop wifi, it's a different story.
```
Small suggestion to change second you are to 'you're': 'but when you're on the go'.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+* Warn users that doing something (e.g. watching video) over cellular could cost them money.
+* Use it in your analytics to gather data on your users’ network quality.
+
+Many applications are already doing something similar. For example, YouTube, Netflix and most other video (or video calling) services automatically adjust the resolution during streaming. When Gmail is loading, it provides users with a link to "load basic HTML (for slow connections)" - a simplified (yet functional) version of Gmail. 
```
Suggestion to remove '- a simplified (yet functional) version of Gmail' phrase as it isn't adding extra value.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+
+<img class="screenshot" src="./loading_gmail_slow_connection.png" alt="A link to load basic HTML version of Gmail when users are on slow connections">
+
+With the Network Information API you can now implement these functionalities in your own applications without much effort.
```
Suggestion to remove this sentence, 'With the'. It's not doing any extra work, especially since you aren't showing how to do it here.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+
+The **navigator.connection** object contains information about a client’s connection. Its properties are:
+
+1. **downlink**: the bandwidth estimate in megabits per second
```
This list would look better in a table.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+
+The **effectiveType** values are also available via [Client Hints](https://www.chromestatus.com/features/5407907378102272) and allow you to communicate the browser's connection type to servers.
+
+<table>
```
I'm not sure I see a lot of value in this table, explaining effectiveType usage. It's not being used here, other than as a descriptor. And it is taking up a lot of space. Would you be open to removing it?

mihajlija replied:
Removed the table from the doc and included a link to the spec.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+The **navigator.connection** object contains information about a client’s connection. Its properties are:
+
+1. **downlink**: the bandwidth estimate in megabits per second
+2. **effectiveType**: the effective type of the connection, with possible values **‘slow-2g’**,**‘2g’**, **‘3g’**, or **‘4g’**
```
I would suggest after 4g having a phrase (covers 4g and higher). That way you don't need an extra full sentence to say just that. I would also suggest moving content in paragraph below that explains how effectiveType is determined up to here.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+
+<img class="screenshot" src="./network_information_in_console.jpg" alt="Chrome DevTools console displaying the values of navigator.connection object's properties">
+
+It might seem odd that the **effectiveType **only goes up to ‘**4g’**, but that segment actually includes all connections that are **‘4g’ and higher.** The API determines the effective connection type based on the combination of **round-trip time (RTT) and downlink speed**. For example, if you have a 10 Mbps downlink combined with 1000 ms latency, the **effectiveType **will be ‘3g’ due to the lat ...
```
As mentioned above, I would suggest moving the description of how effectiveType is determined up to it's definition. I'd also suggest making the language as simple as possible, since it is in a table. Something like: Based on combination of round-trip time and downlink speed. For example, a 10 Mbps downlink combined with 1000ms latency is '3g' due to latency.

Meggin replied:
I'm not entirely sure I understand how effectiveType is calculated, in terms of combinations. Can you have a low RTT and low download link, and if so, how does that combination get calculated? Still isn't quite clear to me-- the calculation aspect of the API.

mihajlija replied:
Rewrote this to "For example, fast downlink combined with high latency will have lower effectiveType due to latency.". Didn't mention any numbers since I've removed the table that references the exact thresholds for calculation. That calculation aspect tripped me up in the beginning too. I was very confused when I would get 3g effectiveType on my laptop at home, but I figured out from the table that it was happening because of the high latency. (That's why I included the table in the first place.)

I think RTT && downlink relationship can be summed up with 'A chain is only as strong as its weakest link'. The worse value of the two will determine the effectiveType.  Low RTT (good) && low downlink (bad) in combination will have the effectiveType that corresponds to the poor downlink value. You'll get '4g' only if both of the values are in the '4g' range from the table.

I hope this helps and please let me know if you think this can be phrased better in the doc :)

addyosmani replied:
I think the rewrite is a lot more clear personally - interested to hear what @Meggin thinks :)

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/index.md:
```diff
@@ -0,0 +1,103 @@
+## Usage
+
+There are many ways you can use this network information to improve the user experience:
+
```
IMPORTANT: This list is interesting to me-- I want to learn about each of them. Is it possible for this guide to focus more on using the network information API to improve experience, rather than covering the what? Maybe even each codelab associated with the doc covers one of these list items?

mihajlija replied:
The thing with this list is that it's highly hypothetical at this point, there aren't (m)any real-world examples I can reference. I could expand on it by writing a couple of sentences about each of these items, but a list feels much more in line with web.dev writing guidelines?

As for the codelabs, some list items, such as warning users about data charges on cellular, are good candidates. Others, such as deferring uploads and downloads might be too complicated for the codelab/under 1000 words format. I'll discuss this with Addy in our weekly call 👍

addyosmani replied:
Hey @Meggin! I think Milica and I would love to see a deeper-dive on NetInfo usecases and code-labs. We'd like to circle back on this later in the year. Right now NetInfo usage is still quite low (we only found one big site using it on HTTP Archive) and I'd love to avoid us investing too deeply just yet on the topic when we have topic migrations we need to get to (like third-party resource loading) :) 

Does that make sense? Maybe we can file an issue to file-up. There's a lot of nuance here like some of these use-cases are quite involved to show a demo of and we'd want to noodle on the best way to present this before forging ahead with codelab content.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+glitch: adaptive-serving-netinfo-starter
+---
+
+# Adaptive serving based on network quality
```
Suggestion to retitle this "Adapt video to image serving based on network quality"

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+---
+page_type: glitch
+title: Adaptive serving based on network quality
```
Said below as well, I'd change this title to be something like "Adapt video to image serving based on network quality'.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+
+# Adaptive serving based on network quality
+
+In this codelab you’ll learn how to adapt your content based on the network quality. This page’s background video should load only when users are on a fast network. On slower networks an image should load instead.
```
Suggestion to remove 'shoulds' and you are missing a comma. Something like: "This page's background video loads only when users are on a fast network. On slower networks, an image loads instead."

mihajlija replied:
Replaced 'should' with 'will' because video to image adaptation isn't implemented at the beginning of the codelab. At this point, _this page's background video loads regardless of the network quality_. It _will load only when users are on a fast network_, but only after people finish the codelab. :)

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+
+# Adaptive serving based on network quality
+
+In this codelab you’ll learn how to adapt your content based on the network quality. This page’s background video should load only when users are on a fast network. On slower networks an image should load instead.
```
Missing comma after 'In this codelab'.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+The [Network Information API](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation) enables you to access information about the user's connection quality. You will use its `effectiveType` property to decide when to serve a video and when to serve an image. `effectiveType` can be ‘slow-2g’, ‘2g’, ‘3g’, or ‘4g’.
+
+## Step 1: Check for browser support
+
```
I'd point out specifically the line in index.html that specifies the video.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
````diff
@@ -0,0 +1,172 @@
+
+Wrap the existing video loading code in an else statement, so that video will still load in browsers that don’t support the Network Information API.
+
+```
````
This first step feels a bit misleading to me. I'd suggest removing it all together, and instead have the 'Check connection type' section cover browser support and connection type. Audience will understand that placeholder code way better than the sort of wiggly presentation of checking browser support without any conditions.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+
+
+## Step 2: Check connection type
+
```
Assuming you've removed check for browser support section above-- you will want to introduce here the two parts of checking connection type-- first checking for browser support, and then checking the connection type. It's a simple change that gets the developer to see what is going on faster.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+
+4. Select **Fast 3G** from throttling presets.
+
+<img class="screenshot" src="./devtools_network_throttling.png" alt='DevTools Network tab with Fast 3G throttling option highlighted'>
```
Image is missing.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+
+The app will update the network information to **3g**:
+
+<img class="screenshot" src="./netinfo_app_3g.png" alt='Matrix-like video background with "NETWORK INFORMATION 3g" text overlay'>
```
Image is missing.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+
+Now reload the page with Fast 3G still enabled and the app will load an image in the background instead of the video: 
+
+<img class="screenshot" src="./netinfo_app_image.png" alt='Matrix-like image background with "NETWORK INFORMATION 3g" text overlay'>
```
Image is missing.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
```diff
@@ -0,0 +1,172 @@
+
+Here’s the final state of the [app on Glitch](https://glitch.com/~adaptive-serving-netinfo).
+
+<img class="screenshot" src="./netinfo_app_video_background.png" alt='Matrix-like video background with "NETWORK INFORMATION 4g" text overlay'>
```
Image is missing.

Meggin commented on content/fast/adaptive-serving-based-on-network-quality/codelab-adaptive-serving-based-on-network-quality.md:
````diff
@@ -0,0 +1,172 @@
+}
+```
+
+## Extra Credit: Respond to changes
````
Suggestion to move extra credit to come after 'try it out' section.

addyosmani commented:
I've previously shared feedback on our calls, so not much to add here. I'd say let's stick with this landing in the `Optimize how you serve content` section, which will include Client Hints and Save Data at a later point. Would be great if we can sort out the image issues Meggin pointed out.

mihajlija commented:
Images are missing due to some issue with the preview server, @robdodson can probably explain better. I've double-checked all of the file paths, I think it will work fine on the staging server. 

I've addressed all of the comments in this rewrite I'm about to push, it's ready for the second round of reviews.

addyosmani commented:
K. Content LGTM. Will defer to @robdodson to see if he might be able to help resolve the file path issues you've been experiencing.

robdodson commented:
Hey @mihajlija,

I think we'll be able to roll our design refresh out next week. It includes a much nicer preview server (yay!) and fancier guides. Would it be ok if we hold off on shipping this guide until that rolls out? I can migrate the guide over and ping you on the PR for review.

mihajlija commented:
Sounds good 👍

Meggin commented:
I can do another content review too in the new design refresh next week.

arthurevans commented:
Hi @mihajlija I'll be a review for @Meggin. @robdodson is the design refresh done? Is this staged somewhere, or can I preview it locally?

robdodson commented:
Hey @arthurevans it looks like Meggin or I need to migrate this post over to the v1 design. I don't think it'll build in the current state.

robdodson commented:
Gonna migrate this. Stay tuned.

robdodson commented:
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

robdodson commented:
Moved to https://github.com/GoogleChrome/web.dev/pull/857
